bpf/lib/nat.h

```diff
@@ -353,6 +353,74 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
 	return DROP_NAT_NO_MAPPING;
 }
 
+static __always_inline int
+snat_v4_rewrite_headers(struct __ctx_buff *ctx, __u8 nexthdr, int l3_off,
+			bool has_l4_header, int l4_off,
+			__be32 old_addr, __be32 new_addr, __u16 addr_off,
+			__be16 old_port, __be16 new_port, __u16 port_off)
+{
+	__wsum sum;
+
+	/* No change needed: */
+	if (old_addr == new_addr && old_port == new_port)
+		return 0;
+
+	sum = csum_diff(&old_addr, 4, &new_addr, 4, 0);
+	if (ctx_store_bytes(ctx, l3_off + addr_off, &new_addr, 4, 0) < 0)
+		return DROP_WRITE_ERROR;
+
+	if (has_l4_header) {
+		int flags = BPF_F_PSEUDO_HDR;
+		struct csum_offset csum = {};
+		__wsum l4_sum = sum;
+
+		csum_l4_offset_and_flags(nexthdr, &csum);
+
+		if (old_port != new_port) {
+			__be32 from, to;
+			int ret;
+
+			switch (nexthdr) {
+			case IPPROTO_TCP:
+			case IPPROTO_UDP:
+				ret = l4_modify_port(ctx, l4_off, port_off,
+						     &csum, new_port, old_port);
+				if (ret < 0)
+					return ret;
+				break;
+#ifdef ENABLE_SCTP
+			case IPPROTO_SCTP:
+				return DROP_CSUM_L4;
+#endif  /* ENABLE_SCTP */
+			case IPPROTO_ICMP:
+				if (l4_store_port(ctx, l4_off, port_off, new_port) < 0)
+					return DROP_WRITE_ERROR;
+
+				from = old_port;
+				to = new_port;
+				l4_sum = csum_diff(&from, 4, &to, 4, 0);
+
+				/* Not initialized by csum_l4_offset_and_flags(): */
+				csum.offset = offsetof(struct icmphdr, checksum);
```

Comment:
Why can't we extend `csum_l4_offset_and_flags`?

Reply:
I think we totally can.

My best guess why it hasn't been done so far is that it introduces overhead to all users. I'd like to understand a bit better how much additional complexity these generic helpers are introducing to paths that weren't exposed to ICMP before (eg. `__lb4_rev_nat()`). And fine-tune if needed. Same would apply here - ideally this would a `ipv4_init_csum()` that doesn't care about ICMPv6.